File: packages/gavio-js/src/interceptors/pii/scanners/phone.ts
/** Phone number scanner — E.164 and common national formats. */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

// E.164 (+CC...) or national groupings with separators. A digit-count filter
// (7–15) below avoids matching short numbers / years.
const PHONE =
  /(?<![\w.])(?:\+?\d{1,3}[ .-]?)?(?:\(\d{1,4}\)[ .-]?)?\d{2,4}(?:[ .-]?\d{2,4}){2,4}(?![\w])/g

export interface PhoneScannerOptions {
  locales?: string[]
}

export function phoneScanner(options: PhoneScannerOptions = {}): PiiScanner {
  const locales = options.locales ?? ['NL', 'DE', 'GB', 'US']
  return {
    entityType: 'PHONE',
    tier: 1,
    confidence: 0.85,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(PHONE)) {
        const digitCount = (m[0].match(/\d/g) ?? []).length
        if (digitCount < 7 || digitCount > 15) continue
        const idx = ctx.nextIndex('PHONE')
        out.push(
          makeMatch({
            entityType: 'PHONE',
            start: m.index,
            end: m.index + m[0].length,
            value: m[0],
            confidence: 0.85,
            replacement: `[PHONE_${idx}]`,
          }),
        )
      }
      return out
    },
    supportsLocale(locale: string): boolean {
      return locales.includes(locale.toUpperCase())
    },
  }
}

File: packages/gavio-py/gavio/interceptors/pii/scanners/phone.py
"""Phone number scanner — E.164 and common national formats."""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

# E.164 (+CC...) or national groupings with separators. Requires at least 7
# digits total to avoid matching short numbers / years.
_PHONE = re.compile(
    r"(?<![\w.])(?:\+?\d{1,3}[ .\-]?)?(?:\(\d{1,4}\)[ .\-]?)?"
    r"\d{2,4}(?:[ .\-]?\d{2,4}){2,4}(?![\w])"
)


class PhoneScanner(PiiScanner):
    entity_type = "PHONE"

    def __init__(self, locales: list[str] | None = None) -> None:
        self.locales = locales or ["NL", "DE", "GB", "US"]

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _PHONE.finditer(text):
            digit_count = sum(c.isdigit() for c in m.group())
            if not 7 <= digit_count <= 15:
                continue
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(),
                    end=m.end(),
                    value=m.group(),
                    confidence=0.85,
                    replacement=f"[PHONE_{idx}]",
                )
            )
        return out

    def supports_locale(self, locale: str) -> bool:
        return locale.upper() in self.locales
